lib/real-device-management.js

```diff
@@ -74,7 +74,9 @@ async function installToRealDevice(device, app, bundleId, opts) {
 
   const {skipUninstall, strategy, timeout} = opts ?? {};
 
-  if (!skipUninstall && bundleId && (await device.isAppInstalled(bundleId))) {
```

Comment:
this looks like a workaround to me. Can the isAppInstalled API be fixed instead to return a proper value?

Reply:
Ideally, yes, I'd love to. I don't have the issue environment to try something, so if i can get such device, I'd love to.


One weird thing was this behavior occurred after uninstalling the app via `ideviceinstaller` for example.
https://github.com/appium/appium-xcuitest-driver/blob/ea9f383ee512cd70170f909f9a13ca2918957406/lib/ios-deploy.js#L181 this filter itself seems correct. The issue happened before https://github.com/appium/appium-xcuitest-driver/pull/1753 , thus probably `lookupApplications` by `usbmuxd` itself returned unexpected data potentially.


The device log also had `Pease try again later` like this error message for the installation failure message, thus potentially it needed more time to complete some internal process.

Reply:
It seems like we need this change.

The "offload" app https://github.com/appium/appium-xcuitest-driver/pull/2050#issuecomment-1740380759 seems not present in a couple of third-party libs as well, but the remove https://github.com/appium/appium-xcuitest-driver/blob/91ec8bf5b6bc68215cc9e05f09f09be3a03c074a/lib/ios-deploy.js#L28 was able to uninstall the app completely from the device even that case.

So, we should address this behavior in docs as well

Reply:
[9dc2288](https://github.com/appium/appium-xcuitest-driver/pull/2050/commits/9dc2288ed6ac0c3fc535827fc88ad4cf0e92d00a) to address this offload app behavior

Reply:
(From another thread to move to correct place)

`skipUninstall` is considered by `isAppInstalled` in `checkAutInstallationState`, so the `(await device.isAppInstalled(bundleId)` removal itself is maybe ok since `skipUninstall` includes the check. It will reduce ~0.2 sec extra time to check `isAppInstalled`